### ISSUE TITLE
traceAddress from debug_traceBlockByNumber is not correct

### ISSUE THREAD

**Issue by ella-quicknode** [ledgerwatch/erigon#9784]
#### System information

Erigon version: `erigon/2.59.0/linux-amd64/go1.21.8"`
OS & Version: `Ubuntu 22.04.4 LTS (GNU/Linux 6.5.0-1019-oracle x86_64)`
Consensus Layer: Lighthouse
Chain/Network: ETH Mainnet 

#### Expected behaviour

Using `debug_traceBlockByNumber` the traceAddress should be consistent to what you see on `trace_block` method. 
`traceAddress` from  `trace_block` method is consistent to Geth trace output. 

Example block `16494085` (0xfbae05) filter with address `0xbaae833f06dac8ef2ab8dea7b707dbcfa048a5f6` 

Geth trace on [etherscan here](https://etherscan.io/vmtrace?txhash=0x536434786ace02697118c44abf2835f188bf79902807c61a523ca3a6200bc350&type=parity#raw). 

#### Actual behaviour

Attaching 2 screen prints on the traceAddress discrepancy for better demostration. 

![trace_block](https://github.com/ledgerwatch/erigon/assets/131174912/aa3d3166-d73c-49df-a734-8e3d7f55bd83)

![debug_traceBlockByNumber](https://github.com/ledgerwatch/erigon/assets/131174912/2457ec32-773f-4df3-a4cd-ca78f023fc16)

#### Steps to reproduce the behaviour

Use this command on the node: 

```
curl -X POST http://localhost:8545 -H 'Content-Type: application/json' -d '{"method":"debug_traceBlockByNumber","params":["0xfbae05", {"tracer": "callTracer"}],"id":1,"jsonrpc":"2.0"}' > response.json
```
I used a script to index the array for trace address.

**Comment by taratorio** [comment]
@ella-quicknode the reason for this is because `trace_block` uses the Parity trace format which does not include CALL/STATICCALLs to precompiles. 

On the other hand, `debug_traceBlockByNumber` uses the native `callTracer` which includes precompiles.

I also checked geth and it follows the same logic, so this does not feel like an issue to me. 

Closing this issue but feel free to re-open if you disagree and/or have a more concrete request.

### More detailed explanation

The reason you are getting [0, 2] in your script instead of [0, 0] is because of these missing calls to precompiles in `trace_block` parity format (as explained above) that happen before the remaining calls (compare the logs in the below 2 sections):
```
Call from 0xef54b1EC1BB9c0F653dB07A6f6409bB3901AC8be to 0x0000000000000000000000000000000000000001: Trace Address = [0 0]
Call from 0xef54b1EC1BB9c0F653dB07A6f6409bB3901AC8be to 0x0000000000000000000000000000000000000001: Trace Address = [0 1]
```

`0x0000000000000000000000000000000000000001` here is a precompiled smart contract implementation of `ecrecover`.

### Calls with `debug_traceBlockByNumber`:
```
Call from 0x519e19Eac9EAd404343c68224dD2A22C5c25F992 to 0xef54b1EC1BB9c0F653dB07A6f6409bB3901AC8be: Trace Address = []
Call from 0xef54b1EC1BB9c0F653dB07A6f6409bB3901AC8be to 0xd9Db270c1B5E3Bd161E8c8503c55cEABeE709552: Trace Address = [0]
Call from 0xef54b1EC1BB9c0F653dB07A6f6409bB3901AC8be to 0x0000000000000000000000000000000000000001: Trace Address = [0 0]
Call from 0xef54b1EC1BB9c0F653dB07A6f6409bB3901AC8be to 0x0000000000000000000000000000000000000001: Trace Address = [0 1]
Call from 0xef54b1EC1BB9c0F653dB07A6f6409bB3901AC8be to 0xBAae833f06daC8ef2ab8DEa7B707DBcfa048a5F6: Trace Address = [0 2]
Call from 0xBAae833f06daC8ef2ab8DEa7B707DBcfa048a5F6 to 0x82ecD135Dce65Fbc6DbdD0e4237E0AF93FFD5038: Trace Address = [0 2 0]
Call from 0xBAae833f06daC8ef2ab8DEa7B707DBcfa048a5F6 to 0x5ef30b9986345249bc32d8928B7ee64DE9435E39: Trace Address = [0 2 0 0]
Call from 0xBAae833f06daC8ef2ab8DEa7B707DBcfa048a5F6 to 0x5ef30b9986345249bc32d8928B7ee64DE9435E39: Trace Address = [0 2 0 1]
Call from 0xBAae833f06daC8ef2ab8DEa7B707DBcfa048a5F6 to 0x5ef30b9986345249bc32d8928B7ee64DE9435E39: Trace Address = [0 2 0 2]
Call from 0xBAae833f06daC8ef2ab8DEa7B707DBcfa048a5F6 to 0x2F0b23f53734252Bda2277357e97e1517d6B042A: Trace Address = [0 2 0 3]
Call from 0xBAae833f06daC8ef2ab8DEa7B707DBcfa048a5F6 to 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2: Trace Address = [0 2 0 4]
Call from 0xBAae833f06daC8ef2ab8DEa7B707DBcfa048a5F6 to 0x2F0b23f53734252Bda2277357e97e1517d6B042A: Trace Address = [0 2 0 5]
Call from 0xBAae833f06daC8ef2ab8DEa7B707DBcfa048a5F6 to 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2: Trace Address = [0 2 0 6]
Call from 0xBAae833f06daC8ef2ab8DEa7B707DBcfa048a5F6 to 0x2F0b23f53734252Bda2277357e97e1517d6B042A: Trace Address = [0 2 0 7]
Call from 0x2F0b23f53734252Bda2277357e97e1517d6B042A to 0x35D1b3F3D7966A1DFe207aa4514C12a259A0492B: Trace Address = [0 2 0 7 0]
Call from 0x2F0b23f53734252Bda2277357e97e1517d6B042A to 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2: Trace Address = [0 2 0 7 1]
Call from 0xBAae833f06daC8ef2ab8DEa7B707DBcfa048a5F6 to 0x19c0976f590D67707E62397C87829d896Dc0f1F1: Trace Address = [0 2 0 8]
Call from 0x19c0976f590D67707E62397C87829d896Dc0f1F1 to 0x35D1b3F3D7966A1DFe207aa4514C12a259A0492B: Trace Address = [0 2 0 8 0]
Call from 0x19c0976f590D67707E62397C87829d896Dc0f1F1 to 0x35D1b3F3D7966A1DFe207aa4514C12a259A0492B: Trace Address = [0 2 0 8 1]
Call from 0xBAae833f06daC8ef2ab8DEa7B707DBcfa048a5F6 to 0x35D1b3F3D7966A1DFe207aa4514C12a259A0492B: Trace Address = [0 2 0 9]
Call from 0xBAae833f06daC8ef2ab8DEa7B707DBcfa048a5F6 to 0x5ef30b9986345249bc32d8928B7ee64DE9435E39: Trace Address = [0 2 0 10]
Call from 0x5ef30b9986345249bc32d8928B7ee64DE9435E39 to 0x35D1b3F3D7966A1DFe207aa4514C12a259A0492B: Trace Address = [0 2 0 10 0]
Call from 0xBAae833f06daC8ef2ab8DEa7B707DBcfa048a5F6 to 0x5ef30b9986345249bc32d8928B7ee64DE9435E39: Trace Address = [0 2 0 11]
Call from 0x5ef30b9986345249bc32d8928B7ee64DE9435E39 to 0x35D1b3F3D7966A1DFe207aa4514C12a259A0492B: Trace Address = [0 2 0 11 0]
Call from 0xBAae833f06daC8ef2ab8DEa7B707DBcfa048a5F6 to 0x35D1b3F3D7966A1DFe207aa4514C12a259A0492B: Trace Address = [0 2 0 12]
Call from 0xBAae833f06daC8ef2ab8DEa7B707DBcfa048a5F6 to 0x9759A6Ac90977b93B58547b4A71c78317f391A28: Trace Address = [0 2 0 13]
Call from 0x9759A6Ac90977b93B58547b4A71c78317f391A28 to 0x35D1b3F3D7966A1DFe207aa4514C12a259A0492B: Trace Address = [0 2 0 13 0]
Call from 0x9759A6Ac90977b93B58547b4A71c78317f391A28 to 0x6B175474E89094C44Da98b954EedeAC495271d0F: Trace Address = [0 2 0 13 1]
```

### Calls with `trace_block`:
```
Call from 0x519e19eac9ead404343c68224dd2a22c5c25f992 to 0xef54b1ec1bb9c0f653db07a6f6409bb3901ac8be: Trace Address = []
Call from 0xef54b1ec1bb9c0f653db07a6f6409bb3901ac8be to 0xd9db270c1b5e3bd161e8c8503c55ceabee709552: Trace Address = [0]
Call from 0xef54b1ec1bb9c0f653db07a6f6409bb3901ac8be to 0xbaae833f06dac8ef2ab8dea7b707dbcfa048a5f6: Trace Address = [0 0]
Call from 0xbaae833f06dac8ef2ab8dea7b707dbcfa048a5f6 to 0x82ecd135dce65fbc6dbdd0e4237e0af93ffd5038: Trace Address = [0 0 0]
Call from 0xbaae833f06dac8ef2ab8dea7b707dbcfa048a5f6 to 0x5ef30b9986345249bc32d8928b7ee64de9435e39: Trace Address = [0 0 0 0]
Call from 0xbaae833f06dac8ef2ab8dea7b707dbcfa048a5f6 to 0x5ef30b9986345249bc32d8928b7ee64de9435e39: Trace Address = [0 0 0 1]
Call from 0xbaae833f06dac8ef2ab8dea7b707dbcfa048a5f6 to 0x5ef30b9986345249bc32d8928b7ee64de9435e39: Trace Address = [0 0 0 2]
Call from 0xbaae833f06dac8ef2ab8dea7b707dbcfa048a5f6 to 0x2f0b23f53734252bda2277357e97e1517d6b042a: Trace Address = [0 0 0 3]
Call from 0xbaae833f06dac8ef2ab8dea7b707dbcfa048a5f6 to 0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2: Trace Address = [0 0 0 4]
Call from 0xbaae833f06dac8ef2ab8dea7b707dbcfa048a5f6 to 0x2f0b23f53734252bda2277357e97e1517d6b042a: Trace Address = [0 0 0 5]
Call from 0xbaae833f06dac8ef2ab8dea7b707dbcfa048a5f6 to 0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2: Trace Address = [0 0 0 6]
Call from 0xbaae833f06dac8ef2ab8dea7b707dbcfa048a5f6 to 0x2f0b23f53734252bda2277357e97e1517d6b042a: Trace Address = [0 0 0 7]
Call from 0x2f0b23f53734252bda2277357e97e1517d6b042a to 0x35d1b3f3d7966a1dfe207aa4514c12a259a0492b: Trace Address = [0 0 0 7 0]
Call from 0x2f0b23f53734252bda2277357e97e1517d6b042a to 0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2: Trace Address = [0 0 0 7 1]
Call from 0xbaae833f06dac8ef2ab8dea7b707dbcfa048a5f6 to 0x19c0976f590d67707e62397c87829d896dc0f1f1: Trace Address = [0 0 0 8]
Call from 0x19c0976f590d67707e62397c87829d896dc0f1f1 to 0x35d1b3f3d7966a1dfe207aa4514c12a259a0492b: Trace Address = [0 0 0 8 0]
Call from 0x19c0976f590d67707e62397c87829d896dc0f1f1 to 0x35d1b3f3d7966a1dfe207aa4514c12a259a0492b: Trace Address = [0 0 0 8 1]
Call from 0xbaae833f06dac8ef2ab8dea7b707dbcfa048a5f6 to 0x35d1b3f3d7966a1dfe207aa4514c12a259a0492b: Trace Address = [0 0 0 9]
Call from 0xbaae833f06dac8ef2ab8dea7b707dbcfa048a5f6 to 0x5ef30b9986345249bc32d8928b7ee64de9435e39: Trace Address = [0 0 0 10]
Call from 0x5ef30b9986345249bc32d8928b7ee64de9435e39 to 0x35d1b3f3d7966a1dfe207aa4514c12a259a0492b: Trace Address = [0 0 0 10 0]
Call from 0xbaae833f06dac8ef2ab8dea7b707dbcfa048a5f6 to 0x5ef30b9986345249bc32d8928b7ee64de9435e39: Trace Address = [0 0 0 11]
Call from 0x5ef30b9986345249bc32d8928b7ee64de9435e39 to 0x35d1b3f3d7966a1dfe207aa4514c12a259a0492b: Trace Address = [0 0 0 11 0]
Call from 0xbaae833f06dac8ef2ab8dea7b707dbcfa048a5f6 to 0x35d1b3f3d7966a1dfe207aa4514c12a259a0492b: Trace Address = [0 0 0 12]
Call from 0xbaae833f06dac8ef2ab8dea7b707dbcfa048a5f6 to 0x9759a6ac90977b93b58547b4a71c78317f391a28: Trace Address = [0 0 0 13]
Call from 0x9759a6ac90977b93b58547b4a71c78317f391a28 to 0x35d1b3f3d7966a1dfe207aa4514c12a259a0492b: Trace Address = [0 0 0 13 0]
Call from 0x9759a6ac90977b93b58547b4a71c78317f391a28 to 0x6b175474e89094c44da98b954eedeac495271d0f: Trace Address = [0 0 0 13 1]
```

**Comment by taratorio** [comment]
@ella-quicknode re-opening this after giving it some more thought. This is not a bug but right now users don't have a way to match the 2 APIs. So what I plan to improve this is to add optional rpc request parameters:
1. "includePrecompiles: <true/false>" to `trace_block` (geth has this option too)
2. "disableDefaultIncludePrecompiles: <true/false>" to `debug_traceBlockByNumber` with `callTracer`

These 2 flags will allow users to pick whichever format they want to stick to and allow them to align the results from the 2 APIs. I'll work on this now and keep you posted.

**Comment by taratorio** [comment]
Above has been merged and will be released in v2.60.3.

There are 2 scenarios to get the 2 results to match:

1. If you are calling `trace_block` with defaults (ie no rpc param inputs) and want to match it to `debug_traceBlockByNumber` then you would need to call the later as follows: 
```
curl http://localhost:8545 \
  -X POST \
  -H "Content-Type: application/json" \
  --data '{"method":"debug_traceBlockByNumber","params":["0xfbae05", {"tracer": "callTracer", "tracerConfig": { "includePrecompiles": false }}],"id":1,"jsonrpc":"2.0"}'
```
<br>
Note: same applies to all other relevant `debug_*` rpc apis like `debug_traceTransaction`, `debug_traceCall`, etc. (interface can be seen [here](https://github.com/ledgerwatch/erigon/blob/release/2.60/turbo/jsonrpc/debug_api.go#L33-L45))

<br>
<br>
2.  If you are calling `debug_traceBlockByNumber` with defaults (ie just with "tracer": "callTracer") and want to match it to `trace_block` then you would need to call the later as follows:
```
curl http://localhost:8545 \
  -X POST \
  -H "Content-Type: application/json" \
  --data '{"method":"trace_block","params":["0xfbae05", null, { "tracerConfig": { "includePrecompiles": true } }],"id":1,"jsonrpc":"2.0"}'
```
<br>
Note: same applies to all other relevant `trace_*` rpc apis like `trace_transaction`, `trace_call`, etc. (interface can be seen [here](https://github.com/ledgerwatch/erigon/blob/release/2.60/turbo/jsonrpc/trace_api.go#L18-L33))
<br>

Hope this is useful. Closing this issue but feel free to re-open if there are still issues.
